internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlySchemaRepository.java

```diff
@@ -128,11 +132,66 @@ private Object doSchemaOperation(String storeName, Function<SchemaData, Object>
     }
   }
 
-  private void mayRegisterAndPopulateMetadataSchema(Store store, SchemaData schemaData) {
+  void maybeRegisterAndPopulateRmdSchema(Store store, SchemaData schemaData) {
     if (store.isActiveActiveReplicationEnabled()) {
       String storeName = store.getName();
-      accessor.subscribeReplicationMetadataSchemaCreationChange(storeName, replicationMetadataSchemaChildListener);
-      accessor.getAllReplicationMetadataSchemas(storeName).forEach(schemaData::addReplicationMetadataSchema);
+      getAccessor().subscribeReplicationMetadataSchemaCreationChange(storeName, replicationMetadataSchemaChildListener);
+      getAccessor().getAllReplicationMetadataSchemas(storeName).forEach(schemaData::addReplicationMetadataSchema);
+    }
+  }
+
+  void maybeRegisterAndPopulateUpdateSchema(Store store, SchemaData schemaData) {
+    if (store.isWriteComputationEnabled()) {
+      String storeName = store.getName();
+      getAccessor().subscribeDerivedSchemaCreationChange(storeName, derivedSchemaChildListener);
+      getAccessor().getAllDerivedSchemas(storeName).forEach(schemaData::addDerivedSchema);
+    }
+  }
+
+  void maybeForceRefreshSchemaDataForSupersetSchemaWithRetry(Store store, SchemaData schemaData, int supersetSchemaId) {
+    if (isSupersetSchemaReadyToServe(store, schemaData, supersetSchemaId)) {
+      return;
+    }
+    RetryUtils.executeWithMaxAttempt(() -> {
+      forceRefreshSchemaData(store, schemaData);
+      if (!isSupersetSchemaReadyToServe(store, schemaData, supersetSchemaId)) {
+        throw new InvalidVeniceSchemaException(
+            "Unable to refresh superset schema id: " + supersetSchemaId + " for store: " + store.getName());
+      }
+    }, 3, Duration.ofMillis(100), Collections.singletonList(InvalidVeniceSchemaException.class));
+
```

Comment:
Instead of adding 2 sets in SchemaData, why not add  the schema ID to a set here, and for next requests, in `getSupersetSchem`  if we find it in the Set, just return the supersetSchemaIdOptional.map(schemaData::getValueSchema);

Reply:
Hmm, I believe we discussed this in the long Slack thread and we agreed either way works. Also single supersetSchemaID set might become invalid when store is changed like non-AA -> AA where RMD schema will need to be checked again. Instead of adding more logic there, a 2-set solution works without more check.

Reply:
yeah, non-AA to AA is one edge case where we need to check both.